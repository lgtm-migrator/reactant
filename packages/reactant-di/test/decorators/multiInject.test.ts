import { injectable, createContainer, multiInject } from '../..';

describe('multiInject', () => {
  test('explicit identifier', () => {
    @injectable()
    class Foo {
      public get test() {
        return 'test';
      }
    }

    @injectable()
    class Bar {
      constructor(@multiInject(Foo) public foos: Foo[]) {}

      public get length() {
        return this.foos.length;
      }
    }

    const bar = createContainer({
      ServiceIdentifiers: new Map(),
      modules: [Foo, Foo],
    }).get(Bar);

    expect(bar.length).toBe(2);
  });

  test('string identifier', () => {
    @injectable()
    class Foo {
      public get test() {
        return 'test';
      }
    }

    @injectable()
    class Bar {
      constructor(@multiInject('Foo') public foos: Foo[]) {}

      public get length() {
        return this.foos.length;
      }
    }

    const bar = createContainer({
      ServiceIdentifiers: new Map(),
      modules: [
        { provide: 'Foo', useClass: Foo },
        { provide: 'Foo', useValue: 'test' },
      ],
    }).get(Bar);

    expect(bar.length).toBe(2);
    expect(bar.foos[1]).toBe('test');
  });

  test('Unexpected multi-Inject', () => {
    @injectable()
    class Foo {
      public get test() {
        return 'test';
      }
    }

    @injectable()
    class Bar {
      constructor(public foos: Foo) {}
    }

    expect(() => {
      createContainer({
        ServiceIdentifiers: new Map(),
        modules: [Foo, Foo],
      }).get(Bar);
    }).toThrowErrorMatchingSnapshot();
  });
});
