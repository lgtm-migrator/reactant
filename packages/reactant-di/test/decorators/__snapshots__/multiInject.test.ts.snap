// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multiInject Unexpected multi-Inject 1`] = `"Ambiguous match found for serviceIdentifier: Foo"`;
